sdks/python/apache_beam/runners/portability/fn_api_runner_transforms.py

```diff
@@ -353,6 +518,382 @@ def make_stage(base_stage, transform):
+  final_stages = frozenset(stages).union(list(replacements.values()))\
+      .difference(list(replacements))
+
+  for stage in final_stages:
+    # Update all references to their final values before throwing
+    # the replacement data away.
+    stage.must_follow = frozenset(replacement(s) for s in stage.must_follow)
+    # Two reads of the same stage may have been fused.  This is unneeded.
+    stage.deduplicate_read()
+  return final_stages
+
+
+def read_to_impulse(stages, pipeline_context):
+  """Translates Read operations into Impulse operations."""
+  for stage in stages:
+    # First map Reads, if any, to Impulse + triggered read op.
+    for transform in list(stage.transforms):
+      if transform.spec.urn == common_urns.deprecated_primitives.READ.urn:
+        read_pc = only_element(transform.outputs.values())
+        read_pc_proto = pipeline_context.components.pcollections[read_pc]
+        impulse_pc = unique_name(
+            pipeline_context.components.pcollections, 'Impulse')
+        pipeline_context.components.pcollections[impulse_pc].CopyFrom(
+            beam_runner_api_pb2.PCollection(
+                unique_name=impulse_pc,
+                coder_id=pipeline_context.bytes_coder_id,
+                windowing_strategy_id=read_pc_proto.windowing_strategy_id,
+                is_bounded=read_pc_proto.is_bounded))
+        stage.transforms.remove(transform)
+        # TODO(robertwb): If this goes multi-process before fn-api
+        # read is default, expand into split + reshuffle + read.
+        stage.transforms.append(
+            beam_runner_api_pb2.PTransform(
+                unique_name=transform.unique_name + '/Impulse',
+                spec=beam_runner_api_pb2.FunctionSpec(
+                    urn=common_urns.primitives.IMPULSE.urn),
+                outputs={'out': impulse_pc}))
+        stage.transforms.append(
+            beam_runner_api_pb2.PTransform(
+                unique_name=transform.unique_name,
+                spec=beam_runner_api_pb2.FunctionSpec(
+                    urn=python_urns.IMPULSE_READ_TRANSFORM,
+                    payload=transform.spec.payload),
+                inputs={'in': impulse_pc},
+                outputs={'out': read_pc}))
+
+    yield stage
+
+
+def impulse_to_input(stages, pipeline_context):
+  """Translates Impulse operations into GRPC reads."""
+  for stage in stages:
+    for transform in list(stage.transforms):
+      if transform.spec.urn == common_urns.primitives.IMPULSE.urn:
+        stage.transforms.remove(transform)
+        stage.transforms.append(
+            beam_runner_api_pb2.PTransform(
+                unique_name=transform.unique_name,
+                spec=beam_runner_api_pb2.FunctionSpec(
+                    urn=bundle_processor.DATA_INPUT_URN,
+                    payload=IMPULSE_BUFFER),
+                outputs=transform.outputs))
+    yield stage
+
+
+def inject_timer_pcollections(stages, pipeline_context):
```

Comment:
Can we add a docstring here for consistency?

Reply:
Done.